teps/0098-workflows.md

```diff
@@ -256,11 +306,164 @@ Why should this TEP _not_ be implemented?
 
 ## Alternatives
 
-<!--
-What other approaches did you consider and why did you rule them out?  These do
-not need to be as detailed as the proposal, but should include enough
-information to express the idea and why it was not acceptable.
--->
+### Build Workflows with Extensible API
+
+In this solution, we'd create a new Workflows API from scratch and support connections to a few commonly used SCMs,
+and secrets stored in Kubernetes secrets. This API would use Triggers as its underlying implementation.
+After creating a working implementation that supports the most common few use cases, we can use this experience to design
+an extensibility mechanism, similar to custom tasks or remote resolution,
+that allows platform builders to support their own event sources and secrets backings.
+Later on, we can declare a Workflows conformance spec if needed.
+
+Pros:
+- Designing for extensibility early on increases the likelihood that we can easily add it later
```

Comment:
This is a good question, I've been thinking we should be more explicit about defining this. The use cases I've identified so far for extensibility are:

- Different VCS providers: e.g. google could support [cloud source repositories](https://cloud.google.com/source-repositories/docs), and there are plenty of other alternatives to GitHub
- Different ways of connecting to the VCS providers we do support: e.g. if a platform builder wants to introduce additional steps for a user connecting their GitHub account. For example, connecting a repo using a platform builder's github app, or introducing a step to link a GitHub repo to a user's account (this idea is admittedly not fully fleshed out but I think @alhuizenga may be thinking about CUJs related to repo connections)
- Easier ways to send notifications to external systems, e.g. designing a mechanism for building slack notification support that doesn't involve sending slack notifications as part of your CI pipeline
- Storing secrets somewhere other than k8s secrets, such as hashicorp vault or google/aws secrets managers. This could potentially be built into pipelines, or maybe these platform builders should just be responsible for implementing CSI secrets store drivers as hashicorp vault has done.
- Storing pipeline data somewhere other than PVCs, like AWS S3. Another idea that we're already exploring building directly into pipelines.

Curious to hear which of these use cases you think are the most important. I'm not sure exactly what you mean about swapping implementation but that isn't a use case I had in mind.

Reply:
Added some more details to the extensibility section to clarify. I think the first two use cases here are definitely in scope (but we don't have to tackle them in the first version of this project), while the last two are better suited for handling in Pipelines.

Reply:
I am happy having any providers in PAC as long it's maintained. If we really need to build an extension mechanism we could have an interface with a service like we have for customtasks in pipeline. I am not sure if we have to draw the line to which git platform we want to support. Do we want to support PaaS platforms like heroku/fly.io and others as well?  Maybe we will need to define a basic set of integrations like webhooks and a web platform with api for  reports.

And as discussed yesterday a lot of the other features is something that probably be better suited for pipeline or even in builtin k8s (secrets/other type of pvc) or for another tekton services like results (notifications/long term storage of logging).

(cloud source repo looks interesting, didn't know about it)